common/tablefunctions.lua

```diff
@@ -230,6 +230,109 @@ if not table.shuffle then
 	end
 end
 
+if not table.map then
+	--- Applies a function to all elements of a table and returns a new table with the results.
+	---@generic K, V, R
+	---@param tbl table<K, V> The input table.
+	---@param callback fun(value: V, key: K, tbl: table<K, V>): R The function to apply to each element. It receives three arguments: the element's value, its key, and the original table.
+	---@return R[] A new table containing the results of applying the callback to each element.
+	function table.map(tbl, callback)
+		local result = {}
+		for k, v in pairs(tbl) do
+			result[k] = callback(v, k, tbl)
+		end
+		return result
+	end
+end
+
+if not table.reduce then
+	--- Reduces a table to a single value by applying a function to each element in order.
+	---@generic K, V, R
+	---@param tbl table<K, V> The input table.
+	---@param callback fun(acc: R, value: V, key: K, tbl: table<K, V>): R The function to apply to each element. It receives four arguments: the accumulator, the element's value, its key, and the original table.
+	---@param initial R The initial value of the accumulator (optional, default: one of the values from the table)
+	---@return R The final value of the accumulator after applying the callback to all elements.
+	function table.reduce(tbl, callback, initial)
+		local accumulator = initial
+
+		for k, v in pairs(tbl) do
+			if initial == nil and accumulator == nil then
+				accumulator = v
```

Comment:
I think this is a bad default because one arbitrary value will exceptionally not have had `callback` called on it and the callback cannot tell which one that is.

If you pass `nil` then the callback can just contain `if acc == nil then ...` to handle it.

Reply:
That sounds reasonable to me. Done.